surl/__init__.py

```diff
@@ -267,22 +305,49 @@ def get_config_from_cli(parser, auth_dir):
     except Exception as e:
         raise CliError("Authorization failed! Double-check password and 2FA. (%s)" % e)
 
-    credentials = json.loads(base64.b64decode(credentials))
-    if credentials["t"] == "macaroon":
-        root = credentials["v"]
+    decoded_credentials = base64.b64decode(credentials)
+    try:
+        credentials = json.loads(decoded_credentials)
+
+        if credentials["t"] == "macaroon":
+            root = credentials["v"]
+            discharge = None
+        elif credentials["t"] == "u1-macaroon":
+            root = credentials["v"]["r"]
+            discharge = credentials["v"]["d"]
+    except json.decoder.JSONDecodeError:
+        # Charmhub just returns the raw credentials, so attempting to parse it fails
+        root = decoded_credentials.decode()
         discharge = None
-    elif credentials["t"] == "u1-macaroon":
-        root = credentials["v"]["r"]
-        discharge = credentials["v"]["d"]
 
     config = ClientConfig(
-        root=root, discharge=discharge, store_env=store_env, path=auth_path
+        root=root, discharge=discharge, store_env=store_env, store_type=store_type, path=auth_path
     )
 
     if auth_path is not None:
         save_config(config)
 
-    return config, remainder
+    return args.url, config, remainder
+
+
+def get_environment_from_url(url):
+    if not url:
+        return "staging", "snapcraft"
+    
+    # The assumption that localhost is SCA can be overriden by a command-line
+    # argument.
+    if "localhost" in url or "127.0.0.1" in url:
+        return "local", "snapcraft"
+    elif "publishergw" in url:
+        return "local", "charmhub"
```

Comment:
both local cases are tricky to detect..what if we just error out if unable to detect?

Reply:
Just looked back at the email conversation I had with Daniel and Przemek, and there it was decided that I'd look for the presence of ":8000" for snapcraft and ":8010" for charmhub; so my first thought is to change this to what was agreed on. I'm also fine with erroring out since I'm keeping the "-s" and added a "-t" option so the cases can be specified manually. @Roadmaster thoughts?

Reply:
ah, looking at just the port sounds a lot more general +1